File: scripts/validate-registry.ts
#!/usr/bin/env node

import { existsSync, readFileSync, statSync } from 'node:fs';
import { join, relative, resolve } from 'node:path';
import { sha256Hex } from '../cli/helper/hash.ts';
import { normalizeLineEndings } from '../cli/helper/line-endings.ts';
import { parseSkillPath } from '../cli/lib.ts';
import { COMBO_SKILLS_MAP, FRONTEND_BONUS_SKILLS, SKILLS_MAP } from '../skills-map.ts';

const __dirname: string = import.meta.dirname;
const PKG_ROOT: string = resolve(__dirname, '..');
const REGISTRY_DIR: string = join(PKG_ROOT, 'skills-registry');
const MANIFEST_PATH: string = join(REGISTRY_DIR, 'index.json');

interface DeclaredSkillInfo {
  full: string;
  sources: Set<string>;
}

interface RegistryEntry {
  files?: string[];
  sha256?: Record<string, string>;
  bundleHash?: string;
  skillPath?: string;
  source?: string;
  commitSha?: string;
  [key: string]: unknown;
}

interface Manifest {
  skills: Record<string, RegistryEntry>;
  [key: string]: unknown;
}

function collectDeclaredSkills(): { declared: Map<string, DeclaredSkillInfo>; conflicts: string[] } {
  const declared = new Map<string, DeclaredSkillInfo>();
  const conflicts: string[] = [];
  const add = (skill: string, source: string): void => {
    const { skillName } = parseSkillPath(skill);
    if (!skillName) return;
    if (!declared.has(skillName)) {
      declared.set(skillName, { full: skill, sources: new Set<string>() });
    } else if (declared.get(skillName)!.full !== skill) {
      conflicts.push(`${skillName}: declared as both ${declared.get(skillName)!.full} and ${skill}`);
    }
    declared.get(skillName)!.sources.add(source);
  };
  for (const tech of SKILLS_MAP) {
    for (const skill of tech.skills) add(skill, tech.id);
  }
  for (const combo of COMBO_SKILLS_MAP) {
    for (const skill of combo.skills) add(skill, combo.id);
  }
  for (const skill of FRONTEND_BONUS_SKILLS) add(skill, 'frontend-bonus');
  return { declared, conflicts };
}

function formatSources(sources: Set<string> | Iterable<string>): string {
  return [...sources].sort().join(', ');
}

function validateEntryFiles(skillName: string, entry: RegistryEntry, errors: string[]): void {
  if (!Array.isArray(entry.files) || entry.files.length === 0) {
    errors.push(`${skillName}: manifest entry has no files`);
    return;
  }
  const shaMap: Record<string, string> = (entry.sha256 ?? {}) as Record<string, string>;
  const parts: string[] = [];
  for (const file of entry.files) {
    const filePath = join(REGISTRY_DIR, skillName, file);
    if (!existsSync(filePath)) {
      errors.push(`${skillName}: missing file ${file}`);
      continue;
    }
    if (!statSync(filePath).isFile()) {
      errors.push(`${skillName}: ${file} is not a file`);
      continue;
    }
    const actualSha = sha256Hex(normalizeLineEndings(readFileSync(filePath)));
    if (shaMap[file] !== actualSha) {
      errors.push(`${skillName}: hash mismatch for ${file}`);
    }
    parts.push(`${file}:${actualSha}`);
  }
  const actualBundleHash = sha256Hex(parts.sort().join('\n'));
  if (entry.bundleHash !== actualBundleHash) {
    errors.push(`${skillName}: bundleHash mismatch`);
  }
}

function main(): void {
  if (!existsSync(MANIFEST_PATH)) {
    throw new Error(`Registry manifest not found: ${MANIFEST_PATH}`);
  }
  const manifest = JSON.parse(readFileSync(MANIFEST_PATH, 'utf-8')) as Manifest;
  const registrySkills: Record<string, RegistryEntry> = (manifest.skills ?? {}) as Record<string, RegistryEntry>;
  const { declared, conflicts } = collectDeclaredSkills();
  const errors: string[] = [];
  for (const conflict of conflicts) errors.push(conflict);
  for (const [skillName, declaredSkill] of declared) {
    const entry = registrySkills[skillName];
    if (!entry) {
      errors.push(
        `${skillName}: declared in skills map (${formatSources(declaredSkill.sources)}) but missing from registry`,
      );
      continue;
    }
    if (entry.skillPath !== declaredSkill.full) {
      errors.push(`${skillName}: registry skillPath is ${entry.skillPath}, expected ${declaredSkill.full}`);
    }
    validateEntryFiles(skillName, entry, errors);
  }
  for (const skillName of Object.keys(registrySkills).sort()) {
    if (!declared.has(skillName)) {
      errors.push(`${skillName}: present in registry but not declared in skills map`);
    }
  }
  if (errors.length > 0) {
    console.error('\nRegistry validation failed:');
    for (const error of errors) console.error(`- ${error}`);
    console.error(`\nChecked manifest: ${relative(process.cwd(), MANIFEST_PATH)}`);
    process.exit(1);
  }
  console.log(
    `Registry validation passed: ${declared.size} declared skill${declared.size === 1 ? '' : 's'} are installable.`,
  );
}

main();

File: src/bin/validate_registry.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::PathBuf;

use skillindex::infra::hash::{normalize_line_endings, sha256_buffer};
use skillindex::registry::parse_skill_path;
use skillindex::skills::{COMBO_SKILLS_MAP, FRONTEND_BONUS_SKILLS, SKILLS_MAP};

fn main() -> anyhow::Result<()> {
    let manifest_dir: PathBuf = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let registry_dir: PathBuf = manifest_dir.join("skills-registry");
    let manifest_path: PathBuf = registry_dir.join("index.json");

    if !manifest_path.exists() {
        anyhow::bail!("Registry manifest not found: {}", manifest_path.display());
    }

    let manifest_str: String = fs::read_to_string(&manifest_path)?;
    let manifest: serde_json::Value = serde_json::from_str(&manifest_str)?;
    let registry_skills: serde_json::Map<String, serde_json::Value> = manifest
        .get("skills")
        .and_then(|v: &serde_json::Value| v.as_object())
        .cloned()
        .unwrap_or_default();

    // Collect declared skills from Rust skills map
    let mut declared: HashMap<String, (String, HashSet<String>)> = HashMap::new();
    let mut conflicts: Vec<String> = Vec::new();

    let mut add = |skill: &str, source: &str| {
        let parsed: skillindex::registry::ParsedSkillPath = parse_skill_path(skill);
        if parsed.skill_name.is_empty() {
            return;
        }
        let entry: &mut (String, HashSet<String>) = declared
            .entry(parsed.skill_name.clone())
            .or_insert_with(|| (skill.to_string(), HashSet::new()));
        if entry.0 != skill {
            conflicts.push(format!(
                "{}: declared as both {} and {}",
                parsed.skill_name, entry.0, skill
            ));
        }
        entry.1.insert(source.to_string());
    };

    for tech in SKILLS_MAP.iter() {
        for &skill in tech.skills {
            add(skill, tech.id);
        }
    }
    for combo in COMBO_SKILLS_MAP.iter() {
        for &skill in combo.skills {
            add(skill, combo.id);
        }
    }
    for &skill in FRONTEND_BONUS_SKILLS {
        add(skill, "frontend-bonus");
    }

    let mut errors: Vec<String> = Vec::new();
    for c in conflicts {
        errors.push(c);
    }

    for (skill_name, (full, sources)) in &declared {
        if let Some(entry) = registry_skills.get(skill_name) {
            let skill_path = entry
                .get("skillPath")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            if skill_path != full {
                errors.push(format!(
                    "{}: registry skillPath is {}, expected {}",
                    skill_name, skill_path, full
                ));
            }
            // Validate files
            let files: Vec<String> = entry
                .get("files")
                .and_then(|v: &serde_json::Value| v.as_array())
                .map(|arr: &Vec<serde_json::Value>| {
                    arr.iter()
                        .filter_map(|v: &serde_json::Value| v.as_str().map(|s: &str| s.to_string()))
                        .collect::<Vec<_>>()
                })
                .unwrap_or_default();
            if files.is_empty() {
                errors.push(format!("{}: manifest entry has no files", skill_name));
                continue;
            }
            let sha_map: HashMap<String, String> = entry
                .get("sha256")
                .and_then(|v| v.as_object())
                .map(|obj: &serde_json::Map<String, serde_json::Value>| {
                    obj.iter()
                        .filter_map(|(k, v)| v.as_str().map(|s: &str| (k.clone(), s.to_string())))
                        .collect::<HashMap<_, _>>()
                })
                .unwrap_or_default();
            let bundle_hash: &str = entry
                .get("bundleHash")
                .and_then(|v: &serde_json::Value| v.as_str())
                .unwrap_or("");

            let mut parts: Vec<String> = Vec::new();
            for file in &files {
                let file_path: PathBuf = registry_dir.join(skill_name).join(file);
                if !file_path.exists() {
                    errors.push(format!("{}: missing file {}", skill_name, file));
                    continue;
                }
                if !file_path.is_file() {
                    errors.push(format!("{}: {} is not a file", skill_name, file));
                    continue;
                }
                let data: Vec<u8> = fs::read(&file_path)?;
                let normalized: Vec<u8> = normalize_line_endings(&data);
                let actual_sha: String = sha256_buffer(&normalized);
                if sha_map.get(file) != Some(&actual_sha) {
                    errors.push(format!("{}: hash mismatch for {}", skill_name, file));
                }
                parts.push(format!("{}:{}", file, actual_sha));
            }
            parts.sort();
            let actual_bundle: String = {
                let joined: String = parts.join("\n");
                sha256_buffer(joined.as_bytes())
            };
            if bundle_hash != actual_bundle {
                errors.push(format!("{}: bundleHash mismatch", skill_name));
            }
        } else {
            let sources_str: String = {
                let mut v: Vec<_> = sources.iter().cloned().collect();
                v.sort();
                v.join(", ")
            };
            errors.push(format!(
                "{}: declared in skills map ({}) but missing from registry",
                skill_name, sources_str
            ));
        }
    }

    for skill_name in {
        let mut v: Vec<_> = registry_skills.keys().cloned().collect();
        v.sort();
        v
    } {
        if !declared.contains_key(&skill_name) {
            errors.push(format!(
                "{}: present in registry but not declared in skills map",
                skill_name
            ));
        }
    }

    if !errors.is_empty() {
        eprintln!("\nRegistry validation failed:");
        for e in &errors {
            eprintln!("- {}", e);
        }
        eprintln!(
            "\nChecked manifest: {}",
            manifest_path
                .strip_prefix(std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")))
                .unwrap_or(&manifest_path)
                .display()
        );
        std::process::exit(1);
    }

    println!(
        "Registry validation passed: {} declared skill{} are installable.",
        declared.len(),
        if declared.len() == 1 { "" } else { "s" }
    );
    Ok(())
}
